navigation/libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/instruction/DefaultInstructionEngine.java

```diff
@@ -0,0 +1,223 @@
+    double userDistance = progress.getCurrentLegProgress().getCurrentStepProgress().getDistanceRemaining();
+    return String.format(
+      Locale.US,
+      CONTINUE_STRING_FORMAT,
+      progress.getCurrentLegProgress().getCurrentStep().getName(),
+      distanceFormatter(userDistance));
+  }
+
+  /**
+   * Creates a {@link String} with the current step distance remaining upcoming step instruction
+   * Example: "In 3.2 miles turn left onto Main St."
+   * @param progress {@link RouteProgress} created by the location change
+   * @return {@link String} with format "In %s %s"
+   */
+  private String buildDefaultFormatInstruction(RouteProgress progress) {
+    double userDistance = progress.getCurrentLegProgress().getCurrentStepProgress().getDistanceRemaining();
+    return String.format(
+      Locale.US,
+      IN_STRING_FORMAT,
+      distanceFormatter(userDistance),
+      convertFirstCharLowercase(progress.getCurrentLegProgress()
+        .getUpComingStep().getManeuver().getInstruction())
+    );
+  }
+
+  /**
+   * Creates a {@link String} with the current step maneuver instruction, current step distance remaining,
+   * and upcoming step instruction
+   * Example: "Turn left onto Main St. then in 3.2 miles turn right onto Second St."
+   * @param progress {@link RouteProgress} created by the location change
+   * @return {@link String} with format "%s then in %s %s"
+   */
+  private String buildThenInFormatInstruction(RouteProgress progress) {
+    double userDistance = progress.getCurrentLegProgress().getCurrentStepProgress().getDistanceRemaining();
+    StepManeuver currentStepManeuver = progress.getCurrentLegProgress().getCurrentStep().getManeuver();
+    return String.format(
+      Locale.US,
+      THEN_IN_STRING_FORMAT,
+      currentStepManeuver.getInstruction(),
+      distanceFormatter(userDistance),
+      convertFirstCharLowercase(progress.getCurrentLegProgress()
+        .getUpComingStep().getManeuver().getInstruction())
+    );
+  }
+
+  /**
+   * Creates a {@link String} with the upcoming step instruction and follow up step instruction
+   * Example: "Turn right onto Main St. then turn left onto Second St."
+   * @param progress {@link RouteProgress} created by the location change
+   * @return {@link String} with format "%s then %s"
+   */
+  private String buildThenFormatInstruction(RouteProgress progress) {
+    int legIndex = progress.getLegIndex();
+    int followUpStepIndex = progress.getCurrentLegProgress().getStepIndex() + 2;
+    return String.format(
+      Locale.US,
+      THEN_STRING_FORMAT,
+      progress.getCurrentLegProgress().getUpComingStep().getManeuver().getInstruction(),
+      convertFirstCharLowercase(progress.getRoute().getLegs().get(legIndex)
+        .getSteps().get(followUpStepIndex).getManeuver().getInstruction())
+    );
+  }
+
+  private static String convertFirstCharLowercase(String instruction) {
+    return instruction.substring(0,1).toLowerCase() + instruction.substring(1);
```

Comment:
You might be checking somewhere else but what happens if the instruction string passed in is null or empty?